### PR TITLE
D: Get the code to compile with all three current d compilers

### PR DESCRIPTION
**Pull request by kuettler**
- Work around a D bug to get exception handling going
- Fix imports
- Modernize pointer downcasts

### CODE REVIEW

File: d/step2_eval.d [hunk 1]
```diff
@@ -16,28 +16,24 @@ MalType READ(string str)
 
 MalType eval_ast(MalType ast, Env env)
 {
-    if (typeid(ast) == typeid(MalSymbol))
+    if (MalSymbol sym = cast(MalSymbol)ast)
```
Comment (dubek) [inline d/step2_eval.d]: should you use `auto sym = ...` here like you did in the other cases?

Comment (kuettler) [reply d/step2_eval.d]: Yes, absolutely. I tried to stick to the original code here. Better to use `auto`.

**Comment by kuettler** [comment]
Thanks a lot for mal. I enjoy it immensely.

I updated the `Dockerfile` to install all three D compilers (gdc and ldc as provided by Ubuntu 18.04 and the latest dmd.) In addition, the `Makefile` has been enhanced to support either of those compilers.

**Comment by kanaka** [comment]
I pushed a docker image to docker hub with the additions and restarted the travis test and D passed. The `apt-get updates` in the Dockerfile aren't necessary (you can do a full rebuild like this `docker build --no-cache -t kanaka/mal-test-d .`). But I'll clean that up myself because I'm also going to add the other two compilers to the travis build. Thanks for the additions!

**Comment by kanaka** [comment]
@kuettler I made the Dockerfile changes I mentioned (plus making dmd available on the path), and I added the linkage to get all three modes available from the top-level: https://github.com/kanaka/mal/commit/c424b18de1a42081417b375daaab12efd8a22e61 All three are now passing in travis: https://travis-ci.org/kanaka/mal/builds/530892861

**Comment by kuettler** [comment]
This is awesome! Thanks for providing mal. Thanks for the D implementation. This is a fantastic way to learn.

**Pull request merged**